tests/test_flatten.py

```diff
@@ -31,3 +45,60 @@ def test_flatten_dict():
             "contextInfo_context2": "info2",
         }
     ]
+
+
+def test_flatten_pd():
+    flattened_pd = flatten(
+        PD_DATA, nested_columns=["contextInfo", "neighborhood"]
+    )
+    assert flattened_pd.equals(
+        DataFrame(
+            [
+                {
+                    "answer": 42,
+                    "contextInfo_context1": "info1",
+                    "contextInfo_context2": "info2",
+                }
+            ]
+        )
+    )
+
+
+def test_flatten_df_is_called():
+    with patch("crossfire._flatten_df", autospec=True) as mock_flatten_df:
```

Comment:
What problem does the `autospec` fixes here? You shared that tests were not passing, but you haven't commented on _why_ they weren't — what was expected? Was there an error? 

I am just trying to understand why we need `autospec` because I have the feeling we don't (e.g. I think there's an error in the return value above).

Reply:
You are rght. I can see that [my comment](https://github.com/FelipeSBarros/crossfire/pull/103#pullrequestreview-1909224735) could be more soecific.

So, basically, even with the `return_value` set correctly, the `contact` function, which is executed in sequence, [fails due inexistence of objects to contatenate](https://github.com/FelipeSBarros/crossfire/actions/runs/8098685427/job/22132667120#step:6:107):

> FAILED tests/test_flatten.py::test_flatten_df_is_called - ValueError: No objects to concatenate

I still not having sure about what the `sutospec` does. So, not sure, what it is fixing...

Reply:
>  I still not having sure about what the sutospec does.

Ok, so **never** add code you don't understand. Never. You can use it to ask: hey, I noticed that when I add that, it works for you. But this is a way to explore the issue rather than a solution.

> No objects to concatenate

So my best guess is that something unexpected happens [when you call the `concat` function](https://github.com/FelipeSBarros/crossfire/blob/f7a01447cd7b722691da7572def59dbc64045d73/crossfire/__init__.py#L84-L90). Have you tried to debug and understand the problem? Try to describe what is happening and why this specific error is popping up.

One idea is to use `import pytest; pytest.set_trace()` just before `concat` and inspect every look of the `for` loop.

Reply:
> Ok, so **never** add code you don't understand. Never. You can use it to ask: hey, I noticed that when I add that, it works for you. But this is a way to explore the issue rather than a solution.

Thanks for your advice!

> So my best guess is that something unexpected happens [when you call the `concat` function](https://github.com/FelipeSBarros/crossfire/blob/f7a01447cd7b722691da7572def59dbc64045d73/crossfire/__init__.py#L84-L90). Have you tried to debug and understand the problem? Try to describe what is happening and why this specific error is popping up.
> 
> One idea is to use `import pytest; pytest.set_trace()` just before `concat` and inspect every look of the `for` loop.

OK, So I will work os that.

Reply:
@cuducos I have been trying to understand the issue with the `test_flatten_df_is_called`, using `pytest.set_trace()`.
using the mock with parameter `autospec` set to `False` (as default). 
The execution of `data.apply(_flatten_df, args=(key,), axis=1)` return the error: "ValueError: No objects to concatenate"

```
poetry run pytest -k test_flatten_df_is_called
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.10.2, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/felipe/repos/crossfire
configfile: pyproject.toml
plugins: ruff-0.2.1, anyio-4.1.0, asyncio-0.21.1
asyncio: mode=strict
collected 94 items / 93 deselected / 1 selected                                                                                                                                          

tests/test_flatten.py 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/felipe/repos/crossfire/crossfire/__init__.py(86)flatten()
-> data = concat(
(Pdb) data.apply(_flatten_df, args=(key,), axis=1)
*** ValueError: No objects to concatenate

# confirming mock 
(Pdb) _flatten_df
<MagicMock name='_flatten_df' id='139870010380432'>
(Pdb) _flatten_df.return_value
contextInfo_context1    info1
contextInfo_context2    info2
dtype: object
```

But running the function without mocking it it returns no ValueError.
I tried to uderstand the issue using chatgpt without sucess until i found [this issues](https://github.com/pandas-dev/pandas/issues/45298) on github.
So, seems to be a problem in the way the mocked function is being used:
> In a context manager, a mocked function in Series.apply calls the function to aggregate instead of the standard one.

Probably, this is why the mock I did is returning ValueErrror related to `concatenate`, as it is being used to `aggregate` instead. It is interesting becuse ocasionally we are using a concatenate function just after its evaluation.

Any way, According to [this comment](https://github.com/pandas-dev/pandas/issues/45298#issuecomment-1009768562) we could overcome the issue by setting `nre_valleble=Mock` in the context manager.
I have tested here and it seems to work well.
```python
def test_flatten_df_is_called():
    with patch("crossfire._flatten_df", new_callable=Mock) as mock_flatten_df:
        mock_flatten_df.return_value = Series(
            {
                "contextInfo_context1": "info1",
                "contextInfo_context2": "info2",
            }
        )

        flatten(PD_DATA, nested_columns=["contextInfo"])

    mock_flatten_df.assert_called_once()
```

But I dont feel confident to commit this code as I couldn't garantee that this was the real issue in the code.

Need your help... :)

Reply:
In cases like that, I like to document the _why_ because the code does not tell the whole story, e.g.:

```python
    # There is a bug on Pandas that makes contact fails with the default MagicMock
    with patch("crossfire._flatten_df", new_callable=Mock) as mock_flatten_df:
```